source/common/http/http2/metadata_decoder.cc

```diff
@@ -1,12 +1,46 @@
 #include "source/common/http/http2/metadata_decoder.h"
 
 #include "source/common/common/assert.h"
+#include "source/common/runtime/runtime_features.h"
 
 #include "absl/container/fixed_array.h"
+#include "quiche/http2/decoder/decode_buffer.h"
+#include "quiche/http2/hpack/decoder/hpack_decoder.h"
+#include "quiche/http2/hpack/decoder/hpack_decoder_listener.h"
 
 namespace Envoy {
 namespace Http {
 namespace Http2 {
+namespace {
+
+class QuicheDecoderListener : public http2::HpackDecoderListener {
+public:
+  explicit QuicheDecoderListener(MetadataMap& map) : map_(map) {}
+  void OnHeaderListStart() override {}
+
+  void OnHeader(const std::string& name, const std::string& value) override {
+    map_.emplace(name, value);
+  }
+
+  void OnHeaderListEnd() override {}
+
+  void OnHeaderErrorDetected(absl::string_view error_message) override {
+    ENVOY_LOG_MISC(error, "Failed to decode payload: {}", error_message);
+    map_.clear();
+  }
+
+private:
+  MetadataMap& map_;
+};
+
+} // anonymous namespace
+
+struct MetadataDecoder::HpackDecoderContext {
```

Comment:
nit: Brief class (well struct) comment please. It might be nice to explain why this struct is helpful instead of just having 2 members of MetadataDecoder.